text/0000-PXE-Booting-UEFI.md

```diff
@@ -0,0 +1,228 @@
+BIOS, UEFI and SecureBoot UEFI. The first two are obvious causing selection of
+the correct DHCP filename when deploying TFTP configuration, the third one
+select "shim.efi" chain loading capability for secure booting.
+
+Similarly to root password, a global setting could be provided for one or both
+flags for missing values (open item).
+
+### [Foreman installer #14920](http://projects.theforeman.org/issues/14920) - Change dhcpd.conf to support EFI loaders
+
+Default DHCPD configuration must be changed in order to allow booting unknown
+UEFI hosts as it was shown above (dhcpd.conf snippet). Using shim.efi is
+perhaps better default than loading Grub directly.
+
+### [Foreman installer #12635](http://projects.theforeman.org/issues/12635) - Options to deploy Grub and PXELinux EFI loaders in TFTP root
+
+It looks like Grub2 is the best and most compatible way of PXE booting EFI
+hardware and it is available in most Linux distributions. It makes sense to
+provide support in our Puppet classes and installer to deploy these.
+Instructions on where to find the firmware in RHEL7 are at
+[access.redhat.com](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Installation_Guide/chap-installation-server-setup.html)
+(RHEL6 does not ship Grub2 so installer should fail or skip that). Similarly
+for Debian and Ubuntu it is available for [direct
+download](http://archive.ubuntu.com/ubuntu/dists/trusty/main/uefi/grub2-amd64/current/)
+or in debian-installer-8-netboot-amd64 package.
+
+UEFI shim shall be installed if possible. Expected file paths are defined in
+the table above.
+
+### [Foreman core #14921](http://projects.theforeman.org/issues/14921) - ComputeResource should pass host loader flag to VM
+
+Since host will have "loader" (BIOS/EFI/SecureBoot EFI) flag, compute resources
+could pass this information when creating new VMs so they gets created with
+correct firmware and PXE booting will just work.
+
+### [Foreman discovery #14922](http://projects.theforeman.org/issues/14922) - Discovery can detect BIOS/EFI and report it back
+
+Discovery can detect current OS loader via a [custom
+fact](https://github.com/jcpunk/puppet-efi/blob/master/lib/facter/efi.rb) and
+this can be used for provisioning to set the "loader" host attribute. Detection
+of secure boot should be possible too, it will be more challenging (depends on
+distribution).
+
+# Other ideas
+[otherideas]: #other-ideas
+
+* UEFI for ARM/ARM64 platform is emerging and it looks like Grub2 will support it - we should test this
+* PPC/PPC64 uses yaboot or petitboot and it should be relatively easy to add support for this but we need someone to test this
+* IA64 uses eLilo - ditto
+* z360 can be booted via zPXE and Cobbler supports that - can we find someone from the community to help with testing that?
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Shall we drop OS variant in favor of pxe template kind?
```

Comment:
To me, operatingsystem@pxe_variant seems like unuseful thing - we only use it with 1:1 mapping with PXE templates:

* syslinux with PXELinux
* pxegrub with PXEGrub
* poap with POAP
* ztp with ZTP

We can simply get rid of pxe_variant, we know the variant thank to the PXE template associated with the OS.